model/speedscope_profile.go

```diff
@@ -114,99 +437,116 @@ func (sp SpeedscopeProfileEvent) appendBeatEvents(cfg *transform.Config, events
 
 	for _, profile := range sp.File.Profiles {
 
-		// Generate a unique profile ID shared by all samples in the
-		// profile.  If we can't generate a UUID for whatever reason,
-		// omit the profile ID.
-		var profileID string
-		if uuid, err := uuid.NewV4(); err == nil {
-			profileID = fmt.Sprintf("%x", uuid)
-		}
+		// XXX only handle eventedProfiles for now
+		if eventedProfile, ok := profile.(*EventedProfile); ok {
 
-		// XXX Duration: profile.go is doing `profileFields["duration"] = pp.Profile.DurationNanos`
-		//     but fields.yml says `Duration of the profile, in microseconds.` What unit is correct?
-		// XXX See "Timestamp troubles" above.
-		//  duration := int64(profile.EndValue - profile.StartValue) ?
-		duration := int64(42)
-
-		var at float64
-		frameIdStack := make(int64Stack, 0) // current stack of indeces into `frames`
-
-		takeSample := func(startVal, endVal float64) {
-			// Generic fields for all events.
-			fields := mapStr{"processor": profileProcessorEntry}
-			if cfg.DataStreams {
-				fields[datastreams.TypeField] = datastreams.MetricsType
-				dataset := fmt.Sprintf("%s.%s", ProfilesDataset, datastreams.NormalizeServiceName(sp.Metadata.Service.Name))
-				fields[datastreams.DatasetField] = dataset
+			// Generate a unique profile ID shared by all samples in the
+			// profile.  If we can't generate a UUID for whatever reason,
+			// omit the profile ID.
+			var profileID string
+			if uuid, err := uuid.NewV4(); err == nil {
+				profileID = fmt.Sprintf("%x", uuid)
 			}
-			sp.Metadata.set(&fields, nil)
 
-			// "profile" field
-			profileFields := common.MapStr{}
-			if profileID != "" {
-				profileFields["id"] = profileID
+			// XXX Duration: profile.go is doing `profileFields["duration"] = pp.Profile.DurationNanos`
+			//     but fields.yml says `Duration of the profile, in microseconds.` What unit is correct?
+			// XXX See "Timestamp troubles" above.
+			//  duration := int64(profile.EndValue - profile.StartValue) ?
+			duration := eventedProfile.EndValue - eventedProfile.StartValue
+			switch eventedProfile.Unit {
```

Comment:
This is skipping the "None" case for now?  (That helps testing with the current simple.speedscope.json at least. :)

Reply:
Intentionally removed the "none" case as the test example uses "none" so wanted to make it pass, but wasn't sure what should be done in the event of "none" for calculating duration, so let it default to assuming the default ingestion unit